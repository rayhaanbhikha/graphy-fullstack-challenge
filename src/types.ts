export interface Coord {
  x: number;
  y: number;
}

export interface AnnotationType {
  id: string;
  coord: Coord;
  text: string;
}

export type PartialAnnotationType = Omit<AnnotationType, 'id'>