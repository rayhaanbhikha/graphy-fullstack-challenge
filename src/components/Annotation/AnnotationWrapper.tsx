import styled from 'styled-components';

import { Coord } from '../../types';

interface IAnnotationWrapper {
  coord: Coord;
}

export const AnnotationWrapper = styled.div<IAnnotationWrapper>`
  position: absolute;
  top: ${props => props.coord.y}px;
  left: ${props => props.coord.x}px;
  display: flex;
`