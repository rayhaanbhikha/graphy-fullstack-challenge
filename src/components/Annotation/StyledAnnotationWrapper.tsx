import styled from 'styled-components';

import { Coord } from '../../types';

interface IStyledAnnotationWrapper {
  coord: Coord;
  isOpen: boolean;
}

export const StyledAnnotationWrapper = styled.div<IStyledAnnotationWrapper>`
  position: absolute;
  top: ${props => props.coord.y}px;
  left: ${props => props.coord.x}px;
  display: flex;
`