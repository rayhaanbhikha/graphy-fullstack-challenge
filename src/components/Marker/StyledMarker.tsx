import styled from 'styled-components';

export const markerDimensions = {
  width: 19,
  height: 19
}

export const StyledMarker = styled.div<{ inEditMode: boolean }>`
  background: rgb(216, 216, 216);
  width: ${markerDimensions.width}px;
  height: ${markerDimensions.height}px;
  border-radius: 3px;
  border: solid 1px rgb(173, 173, 173);
  cursor: pointer;
  z-index: ${props => props.inEditMode ? 999 : 700}
`