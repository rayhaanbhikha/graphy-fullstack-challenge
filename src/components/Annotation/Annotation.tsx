import React, { FunctionComponent, useState } from 'react'

import { StyledMarker } from '../Marker/StyledMarker';
import { AnnotationType } from '../../types';
import { AnnotationTooltip } from '../AnnotationTooltip/AnnotationTooltip';
import { DEFAULT_ID } from '../../Annotations.service';

export interface IAnnotation {
  data: AnnotationType,
}

export const Annotation: FunctionComponent<IAnnotation> = ({ data }) => {
  const isCreatedByUser = data.id === DEFAULT_ID

  const [inEditMode, setinEditMode] = useState(isCreatedByUser);
  const [isHovering, setIsHovering] = useState(isCreatedByUser);

  const onMouseEnter = () => setIsHovering(true)
  const onMouseLeave = () => setIsHovering(false)

  return (
    <StyledMarker
      coord={data.coord}
      inEditMode={inEditMode}
      isHovering={isHovering}
      onClick={(e: React.MouseEvent) => e.stopPropagation()}
      onMouseEnter={onMouseEnter}
      onMouseLeave={onMouseLeave}
    >
      {(isHovering || inEditMode) &&
        <AnnotationTooltip
          data={data}
          inEditMode={inEditMode}
          setinEditMode={setinEditMode}
        />
      }
    </StyledMarker>
  )
}
