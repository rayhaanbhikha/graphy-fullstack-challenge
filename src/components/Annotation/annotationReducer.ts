import { AnnotationState } from "./Annotation";

export enum AnnotationActions {
  SAVE,
  EDIT,
  DELETE,
  OPEN,
  CLOSE,
  DRAG,
}

export const annotationStateReducer = (state: AnnotationState, action: AnnotationActions) => {
  switch (action) {
    case AnnotationActions.SAVE:
      return AnnotationState.OPEN;
    case AnnotationActions.EDIT:
      return AnnotationState.EDITING;
    case AnnotationActions.OPEN:
      if (state === AnnotationState.EDITING) return state;
      return AnnotationState.OPEN;
    case AnnotationActions.CLOSE:
      if (state === AnnotationState.EDITING) return state;
      return AnnotationState.CLOSED;
    case AnnotationActions.DRAG:
      if (state === AnnotationState.EDITING) return state;
      return AnnotationState.DRAGGING;
    case AnnotationActions.DELETE:
    default:
      return state;
  }
}