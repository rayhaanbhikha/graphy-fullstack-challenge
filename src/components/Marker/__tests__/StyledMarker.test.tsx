import React from 'react';
import { render } from '@testing-library/react'
import { AnnotationStates } from '../../Annotation/Annotation';
import { StyledMarker } from '../StyledMarker';
import { Coord } from '../../../types';
import 'jest-styled-components';

describe('StyledMarker', () => {

  const coord: Coord = { x: 5, y: 10 }

  const renderComponent = (state: AnnotationStates) => render(<StyledMarker coord={coord} annotationState={state} />)

  it('should have the default styles', async () => {
    const { asFragment } = await renderComponent(AnnotationStates.EDITING);
    expect(asFragment()).toMatchSnapshot();
  })

  it('should render with correct z-index in edit mode', async () => {
    const { container } = await renderComponent(AnnotationStates.EDITING);
    const element = container.querySelector('div')
    expect(element).toHaveStyleRule('z-index', '900');
  })

  it('should render with correct z-index in open mode', async () => {
    const { container } = await renderComponent(AnnotationStates.OPEN);
    const element = container.querySelector('div')
    expect(element).toHaveStyleRule('z-index', '800');
  })

  it('should render with correct z-index in other mode', async () => {
    const { container } = await renderComponent(AnnotationStates.CLOSED);
    const element = container.querySelector('div')
    expect(element).toHaveStyleRule('z-index', '700');
  })
})
