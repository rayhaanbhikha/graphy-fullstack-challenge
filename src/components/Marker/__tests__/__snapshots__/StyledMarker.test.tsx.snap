// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyledMarker should have the default styles 1`] = `
<DocumentFragment>
  .c0 {
  position: absolute;
  top: 10px;
  left: 5px;
  background: rgb(216,216,216);
  width: 19px;
  height: 19px;
  border-radius: 3px;
  border: solid 1px rgb(173,173,173);
  cursor: pointer;
  z-index: 900;
}

<div
    class="c0"
  />
</DocumentFragment>
`;
